### DISCUSSION TITLE
Legacy skin import does not differentiate between file extensions

### DISCUSSION THREAD

**Discussion by Joppe27** [ppy/osu#21543] [Q&A] [answered]
Edit: the cause of this might not necessarily be the import but could also be `GetTexture` not being able to differentiate between a .png/.wav thus returning `null`. Seems unlikely though, haven't checked the code yet.

Issue happens when importing a legacy skin which has files using the same name but different extensions, for example `spinner-osu.png` and `spinner-osu.wav` which are both used in taiko. At least one of the files (probably both, couldn't test) will fail to import due to them having the same file name making it impossible to use them in-game.

It's very unusual that a skin sprite and sound file have the same file name (this might be the only occurrence?) which is probably why this hasn't been an issue yet.

Skin: https://joppe27.s-ul.eu/vn2hCRBr for stable so you can see what these elements are used for, lazer doesn't use them yet.

**Comment by peppy** [comment]
How are you seeing this manifest in gameplay? Is it always the texture which is missing, but the sound plays correctly?

Initial investigations show that both files *are* correctly imported, so it's likely related to the lookup process if anything.

![Realm Studio 2022-12-07 at 07 02 30](https://user-images.githubusercontent.com/191335/206110753-66a35cfa-3f8c-42dd-a1a9-208aaf70a52b.png)

Also as far as I can tell, the taiko spinner is not implemented in a way that would use the provided sprite/textures, unless I'm missing something..

**Reply by Joppe27** [reply]
I'm working on adding the legacy taiko spinner/swell back (I should've mentioned this, my bad). That's where I noticed GetTexture incorrectly retuning null for only that element specifically. After taking out the .wav file I stopped having issues, so I knew something was going wrong because of the matching file names.

**Reply by peppy** [reply]
Which class or what context are you calling `GetTexture` from? Or rather, what are you calling it against? `ISkinSource`?

**Reply by Joppe27** [reply]
Yes, I'm using ISkinSource.

**Reply by Joppe27** [reply]
I just realized that I can add the file extension to the lookup (even though that's not usually done), which seems to work fine. I probably should've tried this earlier... Is that an okay thing to do to solve this?

**Reply by peppy** [reply]
I don't think that's a good way forward. It looks like this is a valid issue which needs to be fixed at a higher level. I do recall encountering a similar thing recently with skin lookups and don't remember how or if I fixed it.

**Comment by Joppe27** [comment]
Reproduction steps as requested by @bdach:

1. Create a [skin ](https://joppe27.s-ul.eu/IrmrbORS)that includes an image and an audio file. Give both files the same file name (in this example `spinner-osu.png` and `spinner-osu.wav`). 
2. Import the skin and select it in osu!lazer.
3. Attempt to retrieve the image file without its extension using `GetTexture("spinner-osu")`.  You can do this easily by checking out #21601 and deleting the file extension workaround.

- Result: `GetTexture("spinner-osu")` returns `null` (or possibly something else that's not a `Sprite`?)
- Expectation: `GetTexture("spinner-osu")` returns a `Sprite` with `spinner-osu.png` as its content.


Note: I tested this with a `.wav` and `.mp3` extension and both cases produce the issue. I also tried extensions that are not used in osu! such as `.ddd` and `.tiff`, with which the issue did not occur.

**Comment by peppy** [comment, marked as answer]
Tracking via https://github.com/ppy/osu/issues/22084
